bpfd-operator/controllers/bpfd-agent/bpfd-agent.go

```diff
@@ -241,147 +251,141 @@ func (r *BpfProgramReconciler) reconcileBpfProgramConfig(ctx context.Context,
-			controllerutil.RemoveFinalizer(bpfProgram, BpfdAgentFinalizer)
-			err := r.Update(ctx, bpfProgram)
-			if err != nil {
-				return false, err
-			}
-
-			// If K8s hasn't cleaned up here it means we're no longer selected
-			// write NodeNodeSelected status ignoring error (object may not exist)
-			r.updateStatus(ctx, bpfProgram, BpfProgCondNotSelected)
-		}
-
-		return false, nil
-	}
-
-	BpfProgramConfig.Spec.NodeSelector = metav1.LabelSelector{}
-
-	// TODO(astoycos) This will need to end up being a list of loadRequests
-	// if a given BpfProgramConfig selects more than one attach point
-	// (i.e if we support a pod LabelSelector for interfaces) For now
-	// we only support specifying a single BpfProgramAttachPoint in the API so
-	// there will only be a single loadRequest per BpfProgramConfig Object.
-	loadRequest, err := internal.BuildBpfdLoadRequest(BpfProgramConfig)
-	if err != nil {
-		r.updateStatus(ctx, bpfProgram, BpfProgCondNotLoaded)
-
-		return true, fmt.Errorf("failed to generate bpfd load request: %v",
-			err)
-	}
-
-	// Temporary hacks for state which won't match yet based on list API
-	// Proceed-on updates are not currently supported
-	if BpfProgramConfig.Spec.Type == "XDP" || BpfProgramConfig.Spec.Type == "TC" {
-		BpfProgramConfig.Spec.AttachPoint.NetworkMultiAttach.ProceedOn = nil
-		v.Req.AttachPoint.NetworkMultiAttach.ProceedOn = nil
-	}
-
-	// BpfProgram exists but is not correct state, unload and recreate
-	if !reflect.DeepEqual(*v.Req, BpfProgramConfig.Spec) {
-		if retry, err := r.unloadBpfdProgram(ctx, bpfProgram); err != nil {
-			return retry, err
-		}
+			// Re-create correct version
+			return r.loadBpfdProgram(ctx, loadRequest, bpfProgram)
+		} else {
+			// Program already exists, but bpfProgram K8s Object might not be up to date
+			if _, ok := bpfProgram.Spec.Programs[v.Uuid]; !ok {
+				maps, err := internal.GetMapsForUUID(v.Uuid)
+				if err != nil {
+					r.Logger.Error(err, "failed to get bpfProgram's Maps")
+					maps = map[string]string{}
+				}
 
-		// Re-create correct version
-		return r.loadBpfdProgram(ctx, loadRequest, bpfProgram)
-	} else {
-		// Program already exists, but bpfProgram K8s Object might not be up to date
-		if _, ok := bpfProgram.Spec.Programs[v.Uuid]; !ok {
-			maps, err := internal.GetMapsForUUID(v.Uuid)
-			if err != nil {
-				r.Logger.Error(err, "failed to get bpfProgram's Maps")
-				maps = map[string]string{}
-			}
+				bpfProgram.Spec.Programs[v.Uuid] = bpfdiov1alpha1.BpfProgramMeta{
+					AttachPoint: internal.AttachConversion(loadRequest),
+					Maps:        maps,
+				}
 
-			bpfProgram.Spec.Programs[v.Uuid] = bpfdiov1alpha1.BpfProgramMeta{
-				AttachPoint: internal.AttachConversion(loadRequest),
-				Maps:        maps,
-			}
+				r.Logger.V(1).Info("Updating programs from nodestate", "Programs", bpfProgram.Spec.Programs)
+				// Update bpfProgram once successfully loaded
+				if err = r.Update(ctx, bpfProgram, &client.UpdateOptions{}); err != nil {
```

Comment:
Now in one reconcile loop we could end up updating a single bpfProgramConfig object multiple times right? That could definitely break many things

Reply:
Especially since currently we've just implemented the "default interface selector" so the race I'm worried about won't come out now since I  _think_ bpfProgramConfigSpecList should always be one (until we implement podSelector